Cython/Compiler/UtilNodes.py

```diff
@@ -222,13 +233,23 @@ def free_temps(self, code):
 class LetNodeMixin:
     def set_temp_expr(self, lazy_temp):
         self.lazy_temp = lazy_temp
-        self.temp_expression = lazy_temp.expression
+        # don't need to set temp_expression because of property
+
+    # property prevents it from getting out of sync
+    @property
+    def temp_expression(self):
+        return self.lazy_temp.expression
+    @temp_expression.setter
+    def temp_expression(self, value):
+        self.lazy_temp.update_expression(value)
 
     def setup_temp_expr(self, code):
         self.temp_expression.generate_evaluation_code(code)
         self.temp_type = self.temp_expression.type
         if self.temp_type.is_array:
             self.temp_type = c_ptr_type(self.temp_type.base_type)
+        elif self.temp_type.is_cfunction:
```

Comment:
Note to self - this is a workaround for https://github.com/cython/cython/issues/3418. Should be taken out because the fix described in that issue is better.